components/learn/modules/ROOT/pages/on-upgrades.adoc

```diff
@@ -1,8 +1,281 @@
+
+An upgrade then involves the following steps:
+
+. Deploy the new implementation contract.
+. Send a transaction to the proxy, that updates its implementation address to the new one.
+
+NOTE: You can have multiple proxies using the same implementation contract, so you can save gas using this pattern if you plan to deploy multiple copies of the same contract.
+
+Any user of the smart contract always interacts with the proxy, which never changes its address. This allows you to roll out an upgrade or fix a bug without requesting your users to change anything on their end - they just keep interacting with the same address as always.
+
+NOTE: If you want to learn more about how OpenZeppelin proxies work, check out our xref:sdk::pattern.adoc[Upgrades Pattern guide].
+
+[[limitations-of-contract-upgrades]]
+== Limitations of Contract Upgrades
+
+While any smart contract can be made upgradeable, some restrictions of the Solidity language need to be worked around. These come up when writing both the initial contract and version we'll upgrade it to.
+
+=== Initialization
+
+When writing an upgradeable contract, you cannot declare a `constructor`. If you want to run any initialization code, the SDK provides a special `Initializable` base contract that allows you to tag a method as `initializer`, ensuring it can be run only once.
+
+As an example, let's write a new version of the `Box` contract with an initializer, storing the address of an `admin` who will be the only one allowed to change its contents.
+
+```solidity
+// contracts/AdminBox.sol
+pragma solidity ^0.5.0;
+
+import "@openzeppelin/upgrades/contracts/Initializable.sol";
+
+contract AdminBox is Initializable {
+    uint256 private value;
+    address private admin;
+
+    function initialize(address _admin) public initializer {
+        admin = _admin;
+    }
+
+    // Stores a new value in the contract
+    function store(uint256 newValue) public {
+        require(msg.sender == admin);
+        value = newValue;
+        emit ValueChanged(newValue);
+    }
+
+    // Reads the last stored value
+    function retrieve() public view returns (uint256) {
+        return value;
+    }
+}
+```
+
+When deploying this contract, the CLI will prompt us to execute the initializer and ask us to provide the admin address. 
+
+```bash
+$ npx oz create
+✓ Compiled contracts with solc 0.5.9
+? Pick a contract to instantiate: AdminBox
+? Pick a network: development
+✓ Contract AdminBox deployed
+? Call a function to initialize the instance after creating it? Yes
+? Select which function: initialize(_admin: address)
+? _admin (address): 0x90f8bf6a479f320ead074411a4b0e7944ea8c9c1
+✓ Setting everything up to create contract instances
+✓ Instance created at 0x2612Af3A521c2df9EAF28422Ca335b04AdF3ac66
+```
+
+For all practical purposes, the initializer acts as a constructor. However, keep in mind that, since it's a regular function, you need to manually call the initializers of all base contracts, if any.
```

Comment:
```suggestion
For all practical purposes, the initializer acts as a constructor. However, keep in mind that since it's a regular function, you will need to manually call the initializers of all base contracts (if any).
```